desktop-src/SecCNG/creating-a-hash-with-cng.md

```diff
@@ -64,191 +104,226 @@ Abstract:
-                   __in_ecount(argc) LPWSTR *wargv)
-{
+    if (pbHashObject) {
+        HeapFree(GetProcessHeap(), 0, pbHashObject);
+    }
 
-    BCRYPT_ALG_HANDLE       hAlg            = NULL;
-    BCRYPT_HASH_HANDLE      hHash           = NULL;
-    NTSTATUS                status          = STATUS_UNSUCCESSFUL;
-    DWORD                   cbData          = 0,
-                            cbHash          = 0,
-                            cbHashObject    = 0;
-    PBYTE                   pbHashObject    = NULL;
-    PBYTE                   pbHash          = NULL;
+    if (pbHash) {
+        HeapFree(GetProcessHeap(), 0, pbHash);
+    }
+}
+
+int __cdecl wmain(int argc, __in_ecount(argc) LPWSTR* wargv) {
+
+    NTSTATUS status = STATUS_UNSUCCESSFUL;
+    BCRYPT_HASH_HANDLE hHash = NULL;
+    BCRYPT_ALG_HANDLE hAlg = NULL;
+    PBYTE pbHashObject = NULL;
+    DWORD cbHashObject = 0;
+    PBYTE pbHash = NULL;
+    DWORD cbData = 0;
+    DWORD cbHash = 0;
 
     UNREFERENCED_PARAMETER(argc);
     UNREFERENCED_PARAMETER(wargv);
 
     //open an algorithm handle
-    if(!NT_SUCCESS(status = BCryptOpenAlgorithmProvider(
-                                                &hAlg,
-                                                BCRYPT_SHA256_ALGORITHM,
-                                                NULL,
-                                                0)))
-    {
-        wprintf(L"**** Error 0x%x returned by BCryptOpenAlgorithmProvider\n", status);
-        goto Cleanup;
+    status = BCryptOpenAlgorithmProvider(
+        &hAlg, BCRYPT_SHA256_ALGORITHM, NULL, 0
+    );
+
+    if (!NT_SUCCESS(status)) {
+        wprintf(
+            L"**** Error 0x%x returned by BCryptOpenAlgorithmProvider\n",
+            status
+        );
+        wmain_cleanup(hAlg, hHash, pbHashObject, pbHash);
+        return STATUS_UNSUCCESSFUL;
     }
 
     //calculate the size of the buffer to hold the hash object
-    if(!NT_SUCCESS(status = BCryptGetProperty(
-                                        hAlg, 
-                                        BCRYPT_OBJECT_LENGTH, 
-                                        (PBYTE)&cbHashObject, 
-                                        sizeof(DWORD), 
-                                        &cbData, 
-                                        0)))
-    {
+    status = BCryptGetProperty(
+        hAlg,
+        BCRYPT_OBJECT_LENGTH,
+        (PBYTE)&cbHashObject,
+        sizeof(DWORD),
+        &cbData,
+        0
+    );
+
+    if (!NT_SUCCESS(status)) {
         wprintf(L"**** Error 0x%x returned by BCryptGetProperty\n", status);
-        goto Cleanup;
+        wmain_cleanup(hAlg, hHash, pbHashObject, pbHash);
+        return STATUS_UNSUCCESSFUL;
     }
 
     //allocate the hash object on the heap
-    pbHashObject = (PBYTE)HeapAlloc (GetProcessHeap (), 0, cbHashObject);
-    if(NULL == pbHashObject)
-    {
+    pbHashObject = (PBYTE)HeapAlloc(GetProcessHeap(), 0, cbHashObject);
+
+    if (pbHashObject == NULL) {
```

Comment:
This is a regression over the existing code. Counter-intuitive as it may read, the original code would have the compiler fail with an error if the `==` operator were to be replaced by a `=`, by accident.

Contrast that with

```cpp
if (pbHashObject = NULL) { ... }
```

which would issue a warning only, if you're lucky.